flocker/control/test/test_model.py

```diff
@@ -702,3 +720,77 @@ class Record(PRecord):
+        not in the existing ``DeploymentState`` then just add new
+        ``NodeState`` to new ``DeploymentState``.
+        """
+        dataset_id = unicode(uuid4())
+        manifestation = Manifestation(dataset=Dataset(dataset_id=dataset_id),
+                                      primary=True)
+        node = NodeState(
+            hostname=u"node1.example.com",
+            applications={Application(
+                name=u'postgresql-clusterhq',
+                image=DockerImage.from_string(u"image"))},
+            manifestations={dataset_id: manifestation})
+        another_node = NodeState(
+            hostname=u"node2.example.com",
+            applications=frozenset({Application(
+                name=u'site-clusterhq.com',
+                image=DockerImage.from_string(u"image"))}),
+        )
+        original = DeploymentState(nodes=[node])
+        updated = original.update_node(another_node)
+        self.assertEqual((original, updated),
+                         (DeploymentState(nodes=[node]),
+                          DeploymentState(nodes=[node, another_node])))
+
+    def test_update_node_replace(self):
+        """
+        When doing ``update_node()``, if the given ``NodeState`` has hostname
+        in existing ``DeploymentState`` node then update all non-``None``
+        attributes that ``NodeState`` in the new ``Deployment``.
+        """
+        node = NodeState(
+            hostname=u"node1.example.com",
+            applications=[])
+
+        update_applications = NodeState(
+            hostname=u"node1.example.com",
+            applications=frozenset({Application(
+                name=u'site-clusterhq.com',
+                image=DockerImage.from_string(u"image"))}),
+            used_ports=[1, 2],
+            manifestations=None,
+            paths=None,
+        )
+        dataset_id = unicode(uuid4())
+        manifestation = Manifestation(dataset=Dataset(dataset_id=dataset_id),
+                                      primary=True)
+        update_manifestations = NodeState(
+            hostname=u"node1.example.com",
+            applications=None,
+            used_ports=None,
+            paths={dataset_id: FilePath(b"/xxx")},
+            manifestations={dataset_id: manifestation})
+
+        original = DeploymentState(nodes=[node])
+        updated = original.update_node(update_applications).update_node(
+            update_manifestations)
+        self.assertEqual(updated,
+                         DeploymentState(nodes=[NodeState(
+                             hostname=u"node1.example.com",
+                             applications=frozenset({Application(
+                                 name=u'site-clusterhq.com',
+                                 image=DockerImage.from_string(u"image"))}),
+                             used_ports=[1, 2],
+                             paths={dataset_id: FilePath(b"/xxx")},
+                             manifestations={dataset_id: manifestation})]))
```

Comment:
Some duplication could be avoided by defining this expected `DeploymentState` first and then defining `update_applications` and `update_manifestations` in terms of it (eg `update_applications = expected.set(manifestations=None, paths=None)`.